### ISSUE TITLE
Concatenation of ListSets does not preserve ordering

### ISSUE THREAD

**Issue by bpiper** [scala/bug#12316]
## reproduction steps

Scala version 2.12.12 (works in 2.12.10 and 2.13.x)

```
Welcome to Scala 2.12.12 (OpenJDK 64-Bit Server VM, Java 11.0.8).
Type in expressions for evaluation. Or try :help.

scala> import scala.collection.immutable.ListSet
import scala.collection.immutable.ListSet

scala> val foo = ListSet(1)
foo: scala.collection.immutable.ListSet[Int] = ListSet(1)

scala> val bar = foo ++ ListSet(1, 2, 3)
bar: scala.collection.immutable.ListSet[Int] = ListSet(1, 3, 2)
```

This was previously working as expected in 2.12.10.

```
Welcome to Scala 2.12.10 (OpenJDK 64-Bit Server VM, Java 11.0.8).
Type in expressions for evaluation. Or try :help.

scala> import scala.collection.immutable.ListSet
import scala.collection.immutable.ListSet

scala> val foo = ListSet(1)
foo: scala.collection.immutable.ListSet[Int] = ListSet(1)

scala> val bar = foo ++ ListSet(1, 2, 3)
bar: scala.collection.immutable.ListSet[Int] = ListSet(1, 2, 3)
```

## problem

Order of elements should be preserved (as it is in 2.12.10 and earlier, and in 2.13.x).

Workaround is to first convert the argument to a traversable other than ListSet, e.g.

```
scala> val baz = foo ++ ListSet(1, 2, 3).toVector
baz: scala.collection.immutable.ListSet[Int] = ListSet(1, 2, 3)
```

**Comment by som-snytt** [comment]
The scaladoc says elements are stored in reverse order, and the last commit special cases `ListSet` for efficiency and structural sharing. Probably something has to be reversed.
```
scala> ListSet(1,2,3,42,43,44) ++ ListSet(42,43,44,27,28,29)
res1: scala.collection.immutable.ListSet[Int] = ListSet(1, 2, 3, 42, 43, 44, 29, 28, 27)

scala> ListSet(1,2,3,42,43,44) ++ ListSet(10,11,12,42,43,44,27,28,29)
res3: scala.collection.immutable.ListSet[Int] = ListSet(1, 2, 3, 42, 43, 44, 29, 28, 27, 12, 11, 10)
```
I wonder if the new collections group has a kind of Batsignal.

**Comment by SethTisue** [comment]
if only we could just rub a magic lamp labeled @scala/collections and the squad would appear! we can only wish such sorcery were possible

I am closing the ticket because the problem isn't reproducible in 2.13, but we are still doing 2.12.x releases and accepting 2.12-only fixes.

I suppose this would have regressed in https://github.com/scala/scala/pull/8589 ? cc @mkeskells

**Comment by bpiper** [comment]
@SethTisue what's the best way to get this fixed under 2.12.x? Just submit a PR mentioning this issue?

**Comment by mkeskells** [comment]
will have a look

**Comment by som-snytt** [comment]
I guess the batsignal is the `@` at signal. @bpiper it seems Seth uses the at signal when no one answers the at phone.

Yes, ordinarily you could beat the superhero to the punch, so to speak.

**Comment by NthPortal** [comment]
looks like the fix should be a simple type test added to `ListSet` to handle concatenation specially

@SethTisue shouldn't the milestone be 2.12.14?

**Comment by SethTisue** [comment]
> shouldn't the milestone be 2.12.14?

I see what you mean, but my reasoning is that 2.13.0 is the earliest released version in which the problem doesn't exist.

> what's the best way to get this fixed under 2.12.x? Just submit a PR mentioning this issue?

@bpiper indeed, but maybe coordinate with @mkeskells
